aten/src/ATen/native/sparse/SparseTensor.cpp

```diff
@@ -101,7 +100,16 @@ SparseTensor new_with_dims_and_tensor_sparse(
     const TensorOptions& options) {
   SparseTensor self = new_sparse(options);
   get_sparse_impl(self)->resize_(sparse_dim, dense_dim, size);
-  alias_into_sparse(self, indices, values);
+  // NOTE: There is no guarantee that `indices` and `values` don't contain AutogradMeta. However,
+  // we want to maintain the invariant that `indices_` and `values_` of a sparse tensor don't
+  // contain AutogradMeta, and to achieve that we shallow-copy `indices` and `values` here.
+  auto indices_shallow_copy = LongTensor(indices.unsafeGetTensorImpl()->shallow_copy_and_detach(
```

Comment:
I'm a little confused about what is going on with the version counter.

Given this change, I'd expect that when I call `SparseTensor._values()` I get a tensor that shares the version counter with the values that were originally passed in.  But that doesn't seem to be the case, either in this PR or previously -- do you know why this is?

Example:
```
>>> ind=torch.tensor([[0],[1]]).add_(1).sub_(1)
>>> values = torch.tensor([1.]).add_(1).add_(1).sub_(1).sub_(1)
>>> c=torch.sparse_coo_tensor(ind, values)
>>> c._values()._version
0
>>> c._indices()._version
0
>>> c=torch.sparse_coo_tensor(ind, values).coalesce()
>>> c.values()._version
0
>>> c.indices()._version
0
```

Reply:
The current implementation of `VariableType::_values()` makes it so that `c._values()` shares the same version counter as `c`. I think this is the right semantics because updating `c._values()` should also bump the version counter of `c`.

A related issue is we need to make sure in-place update on the original value tensor `values` also updates the version counter of the sparse tensor's `c`'s `values_` tensor, and it should throw version mismatch error in the `backward()` operation if the original value tensor is changed (this requires saving the value tensor for backward in the sparse constructor). I added the task in https://github.com/pytorch/pytorch/issues/13638.